File: typescript/Structural/09-Decorator/Decorator/InsuredDecorator.ts
import { IJobCard } from "../Component/IJobCard";
import { JobCardDecorator } from "./JobCardDecorator";

export class InsuredDecorator extends JobCardDecorator {
  private static readonly INSURANCE_PREMIUM = 350;

  constructor(jobCard: IJobCard) {
    super(jobCard);
  }

  get cost(): number {
    return this.wrapped.cost + InsuredDecorator.INSURANCE_PREMIUM;
  }

  get description(): string {
    return `🛡️  INSURED | ${this.wrapped.description}`;
  }

  execute(): void {
    console.log("  📄 Generating liability waiver...");
    super.execute();
    console.log("  ✅ Insurance certificate issued.");
  }
}

File: csharp/Structural/09-Decorator/Decorator/InsuredDecorator.cs
using Decorator.Component;

namespace Decorator.Decorator;

public class InsuredDecorator : JobCardDecorator
{
    private const decimal InsurancePremium = 350m;

    public InsuredDecorator(IJobCard jobCard) : base(jobCard) { }

    public override decimal Cost => _wrapped.Cost + InsurancePremium;
    public override string Description => $"🛡️  INSURED | {_wrapped.Description}";

    public override void Execute()
    {
        Console.WriteLine("  📄 Generating liability waiver...");
        base.Execute();
        Console.WriteLine("  ✅ Insurance certificate issued.");
    }
}
